blog/2022-04-12-ephemeral-containers.md

```diff
@@ -0,0 +1,183 @@
+
+### OPA Gatekeeper
+
+[OPA Gatekeeper](https://github.com/open-policy-agent/gatekeeper) does not require any code changes as all of its policies are written in [rego](https://www.openpolicyagent.org/docs/latest/policy-language/). It's sub project [Gateekper Library](https://github.com/open-policy-agent/gatekeeper-library/) does however have to be updated. The library contains an implementation of the common Pod Security Policies. This includes policies like not allowing containers in privileged mode. The issue with the all of the policies is that they currently only check containers specified in `initContainers` and `containers`, analyze the [following](https://github.com/open-policy-agent/gatekeeper-library/blob/275a1628694dcdf9daf5f6dda1373de6af78e7da/library/pod-security-policy/privileged-containers/template.yaml#L49-L55) rego as an example.
+
+The good news is that this is a pretty easy fix, the bad news is that it requires end users to update the policies pulled from the library.
+
+### Kyverno
+
+[Kyverno](https://github.com/kyverno/kyverno) seems to have resolved the issues faster. Compared to OPA Gatekeeper however it did require a small code change which means that version [1.5.3](https://github.com/kyverno/kyverno/releases/tag/v1.5.3) or later is needed to write policies for ephemeral containers. They have also [updated their policy library](https://github.com/kyverno/policies/pull/241) to include checking ephemeral containers. Kyverno has done a great job solving these issues quickly. It does still require end users to update however.
+
+### Pod Security Policies
+
+[Pod Security Policies](https://kubernetes.io/docs/concepts/security/pod-security-policy/) used to be the default policy tool for Kubernetes, and a lot of projects have rules based on Pod Security Policies (PSP). However if you are relying on PSP in a modern cluster you should really start looking for other options like OPA Gatekeeper or Kyverno. PSP has been deprecated since Kubernetes v1.21 and will be removed in v1.25.
+
+If PSP is your only policy tool and you are planning to upgrade to v1.23, don't. As PSP is deprecated no new features have been added, and that includes policy enforcement on ephemeral containers. Which means that any security context in an ephemeral container is allowed no matter the PSP in the cluster. 
```

Comment:
Would it be good to mention the new version of psp is coming? https://kubernetes.io/docs/concepts/security/pod-security-admission/

Reply:
Was not aware of that work. Is it worth mentioning if I havent verified that it actually blocks?

Reply:
I think it's enough to just mention that it's coming and if you are a psp user it's something that is worth looking into.